### PR TITLE
Finagle issue 343 adjuster

### PR DESCRIPTION
**Pull request by naoman**
Adding adjuster to handle [finagle issue 343](https://github.com/twitter/finagle/issues/343). This adjuster handles a bug in finagle memcached library where the duration of a span ends up being more than the actual value because a span is submitted twice.

The fix is to drops finagle.flush annotation for memcache spans. We look for "Hit" or "Miss" binary annotations to make sure the span is coming from memcache.

### CODE REVIEW

File: adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    Builder adjustIssue343(boolean adjustIssue343);
+
+    FinagleIssue343Adjuster build();
+  }
+
+  @Override protected boolean shouldAdjust(Span span) {
+    if (adjustIssue343()) {
+      boolean containsFinagleFlushAnnotation = false;
+      for (Annotation a : span.annotations) {
```
Comment (mansu) [inline adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java]: Please consider making this check into a static method.

File: adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        if (a.value.equals("finagle.flush")) {
+          containsFinagleFlushAnnotation = true;
+          break;
+        }
+      }
+      if (containsFinagleFlushAnnotation) {
+        for (BinaryAnnotation b : span.binaryAnnotations) {
```
Comment (mansu) [inline adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java]: Please consider refactoring this into a static method also. Then this method becomes 
return checkContainsFinagleFlushAnnotation(span) && isMemcacheHit(span);

File: adjuster/finagle/src/test/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343AdjusterTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      .build();
+
+  Span serverSpanWithFinagleFlushWithoutHitBinaryAnnotation =
+      serverSpanWithFinagleFlush.toBuilder()
+          .binaryAnnotations(new ArrayList<>())
+          .build();
+
+  /** Default should not do any adjustment */
+  @Test
+  public void adjusterDisabled() {
```
Comment (mansu) [inline adjuster/finagle/src/test/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343AdjusterTest.java]: nice tests!

**Comment by naoman** [comment]
#26

File: adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * The fix is to drops finagle.flush annotation for memcache spans. We look for "Hit" or "Miss"
+     * binary annotations to make sure the span is coming from memcache.
+     *
+     * For more details see
+     * https://github.com/twitter/finagle/issues/343
+     *
+     * Default false
+     */
+    Builder adjustIssue343(boolean adjustIssue343);
```
Comment (adriancole) [inline adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java]: seems this condition should be outside this class..

File: autoconfigure/adjuster-finagle/README.md [hunk 5]
```diff
@@ -34,3 +34,4 @@ have a prefix of "zipkin.sparkstreaming.adjuster.finagle"
 Property | Default | Description | Fix
 --- | --- | --- | ---
 apply-timestamp-and-duration | true | Backfill span.timestamp and duration based on annotations. | [Use zipkin-finagle](https://github.com/openzipkin/zipkin-finagle/issues/10)
+adjust-issue343 | false | Drops "finagle.flush" annotation, to rectify [finagle memcached bug](https://github.com/twitter/finagle/issues/343). | Used finagle version 6.36.0 or higher
```
Comment (adriancole) [inline autoconfigure/adjuster-finagle/README.md]: s/used/use

File: adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Contains adjustments for finagle issue 343 (https://github.com/twitter/finagle/issues/343).
+ * If applied, this adjuster drops finagle.flush annotation.
+ */
+@AutoValue
+public abstract class FinagleIssue343Adjuster extends Adjuster{
```
Comment (adriancole) [inline adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java]: If it were me, I'd make this public final and then with a package-protected constructor. Add to that a factory method like `public static FinagleIssue343Adjuster create()` which calls the package protected constructor.

File: adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  public static FinagleIssue343Adjuster.Builder newBuilder() {
+    return new AutoValue_FinagleIssue343Adjuster.Builder();
+  }
+
+  @AutoValue.Builder
+  public interface Builder {
+    /**
```
Comment (adriancole) [inline adjuster/finagle/src/main/java/zipkin/sparkstreaming/adjuster/finagle/FinagleIssue343Adjuster.java]: I think this more detailed note will be read more often if on the type.. it isn't specific to the build method.

**Review by codefromthecrypt — APPROVED** [review]
looks good! some nits which you can choose to do or not do. merge when you're ready

**Pull request merged**